File: enter.pollinations.ai/scripts/user-pipeline/daily-spore-recheck.ts
#!/usr/bin/env npx tsx
/**
 * Daily spore recheck pipeline.
 *
 * Scores 1/7 of all spore users per day via GitHub activity (age, repos, commits, stars).
 * Users above the threshold are promoted to seed; deleted accounts are banned.
 * Runs until all spores have been checked (takes ~7 days for a full cycle).
 *
 * Usage:
 *   cd enter.pollinations.ai
 *   npx tsx scripts/user-pipeline/daily-spore-recheck.ts
 *   npx tsx scripts/user-pipeline/daily-spore-recheck.ts --dry-run
 *   npx tsx scripts/user-pipeline/daily-spore-recheck.ts --emails-file /tmp/emails.txt
 *
 * Options:
 *   --dry-run        Preview actions without writing to D1
 *   --verbose / -v   Print per-user score breakdown
 *   --emails-file    Restrict to emails in a newline-separated file
 */

import { TIER_POLLEN } from "../../src/tier-config.ts";
import {
    extractDeletedGithubIds,
    type GitHubValidationResult,
    isScorableValidationResult,
    storeGithubScores,
    validateUserRecords,
} from "./scoring/github-score.ts";
import { executeD1, queryD1 } from "./shared/d1.ts";
import { buildEmailFilter, loadEmailCohort } from "./shared/email-cohort.ts";
import {
    banUsersByEmails,
    banUsersByGithubIds,
    GITHUB_ID_INVALID_REASON,
    PIPELINE_DB_BATCH_SIZE,
} from "./shared/github-identity.ts";

type Environment = "staging";

interface ParsedArgs {
    env: Environment;
    dryRun: boolean;
    verbose: boolean;
    cohortEmails: string[] | null;
}

interface SporeRow {
    email: string;
    github_id: number | null;
}

const MAX_USERS_PER_RUN = 8000;

function parseArguments(): ParsedArgs {
    const args = process.argv.slice(2);
    const envIndex = args.indexOf("--env");
    const env =
        envIndex >= 0 && args[envIndex + 1] ? args[envIndex + 1] : "staging";
    const emailsFileIndex = args.indexOf("--emails-file");
    const emailsFile =
        emailsFileIndex >= 0 && args[emailsFileIndex + 1]
            ? args[emailsFileIndex + 1]
            : undefined;

    if (env !== "staging") {
        console.error(
            `❌ Unsupported --env ${env}. This branch is locked to staging and cannot write to production.`,
        );
        process.exit(1);
    }

    let cohortEmails: string[] | null = null;
    try {
        cohortEmails = loadEmailCohort(emailsFile);
    } catch (error) {
        console.error(
            `❌ ${error instanceof Error ? error.message : String(error)}`,
        );
        process.exit(1);
    }

    return {
        env: "staging",
        dryRun: args.includes("--dry-run"),
        verbose: args.includes("--verbose") || args.includes("-v"),
        cohortEmails,
    };
}

function fetchSporeCount(
    env: Environment,
    cohortEmails: string[] | null,
): number {
    const emailFilter = buildEmailFilter("email", cohortEmails);
    const rows = queryD1(
        env,
        `SELECT COUNT(*) AS count FROM user WHERE tier = 'spore' AND COALESCE(banned, 0) = 0${emailFilter}`,
    );
    return Number(rows[0]?.count ?? 0);
}

function fetchSporeSlice(
    env: Environment,
    cohortEmails: string[] | null,
): { rows: SporeRow[]; totalSpores: number; sliceSize: number } {
    const totalSpores = fetchSporeCount(env, cohortEmails);
    if (totalSpores === 0) {
        return { rows: [], totalSpores: 0, sliceSize: 0 };
    }

    const sliceSize = Math.min(Math.ceil(totalSpores / 7), MAX_USERS_PER_RUN);
    const emailFilter = buildEmailFilter("email", cohortEmails);
    const rows = queryD1(
        env,
        `SELECT email, github_id FROM user WHERE tier = 'spore' AND COALESCE(banned, 0) = 0${emailFilter} ORDER BY score_checked_at ASC, created_at ASC, email ASC LIMIT ${sliceSize}`,
    ) as SporeRow[];

    return { rows, totalSpores, sliceSize };
}

function extractRiskBlockedGithubIds(
    results: GitHubValidationResult[],
): number[] {
    return Array.from(
        new Set(
            results.flatMap((result) =>
                result.risk_status === "suspicious" &&
                Number.isInteger(result.github_id) &&
                result.github_id > 0
                    ? [result.github_id]
                    : [],
            ),
        ),
    );
}

function upgradeUsers(env: Environment, githubIds: number[]): number {
    const uniqueIds = Array.from(
        new Set(
            githubIds.filter(
                (githubId): githubId is number =>
                    Number.isInteger(githubId) && githubId > 0,
            ),
        ),
    );
    let upgraded = 0;

    for (
        let index = 0;
        index < uniqueIds.length;
        index += PIPELINE_DB_BATCH_SIZE
    ) {
        const batch = uniqueIds.slice(index, index + PIPELINE_DB_BATCH_SIZE);
        if (batch.length === 0) continue;

        const ok = executeD1(
            env,
            `UPDATE user SET tier = 'seed', tier_balance = ${TIER_POLLEN.seed} WHERE github_id IN (${batch.join(", ")}) AND tier = 'spore'`,
        );
        if (ok) upgraded += batch.length;
    }

    return upgraded;
}

function summarize(results: GitHubValidationResult[]): void {
    const approved = results.filter((result) => result.approved);
    const rejected = results.filter((result) => !result.approved);
    const suspicious = results.filter(
        (result) => result.risk_status === "suspicious",
    );

    console.log("\n📊 Validation summary:");
    console.log(`   Approved by score: ${approved.length}`);
    console.log(`   Rejected by score: ${rejected.length}`);
    console.log(`   Suspicious GitHub profiles: ${suspicious.length}`);

    if (results.length > 0) {
        const average =
            results.reduce(
                (sum, result) => sum + Number(result.details?.total ?? 0),
                0,
            ) / results.length;
        console.log(`   Average score: ${average.toFixed(2)}`);
    }
}

async function main(): Promise<void> {
    const config = parseArguments();
    const { rows, totalSpores, sliceSize } = fetchSporeSlice(
        config.env,
        config.cohortEmails,
    );

    console.log("🌱 Daily Spore Recheck");
    console.log(`   Environment: ${config.env}`);
    console.log(`   Mode: ${config.dryRun ? "DRY RUN" : "LIVE"}`);
    if (config.cohortEmails) {
        console.log(`   Email cohort: ${config.cohortEmails.length} users`);
    }
    console.log(`   Total spores: ${totalSpores}`);
    console.log(`   Daily target: ${sliceSize}`);
    console.log(`   Selected users: ${rows.length}`);

    if (rows.length === 0) {
        console.log("✅ No spore users to process");
        return;
    }

    const invalidRows = rows.filter(
        (row) => !Number.isInteger(row.github_id) || row.github_id === null,
    );
    const validRows = rows.filter(
        (row): row is SporeRow & { github_id: number } =>
            Number.isInteger(row.github_id) && row.github_id !== null,
    );

    if (invalidRows.length > 0) {
        if (config.dryRun) {
            console.log(
                `🚫 Dry run would ban ${invalidRows.length} spore users with missing/invalid GitHub IDs`,
            );
        } else {
            const banned = banUsersByEmails(
                config.env,
                invalidRows.map((row) => row.email),
                GITHUB_ID_INVALID_REASON,
            );
            console.log(
                `🚫 Banned ${banned} spore users with missing/invalid GitHub IDs`,
            );
        }
    }

    if (validRows.length === 0) {
        console.log("✅ No valid spore users left for GitHub scoring");
        return;
    }

    const results = await validateUserRecords(validRows);
    const resultsByGithubId = new Map(
        results.flatMap((result) =>
            Number.isInteger(result.github_id) && result.github_id > 0
                ? [[result.github_id, result] as const]
                : [],
        ),
    );
    const orderedResults = validRows.flatMap((row) => {
        const result = resultsByGithubId.get(row.github_id);
        return result ? [result] : [];
    });
    const deletedGithubIds = extractDeletedGithubIds(orderedResults);
    const scoreableResults = orderedResults.filter(isScorableValidationResult);
    const unavailableCount =
        orderedResults.length -
        deletedGithubIds.length -
        scoreableResults.length;
    const riskBlockedGithubIds = extractRiskBlockedGithubIds(scoreableResults);
    const riskBlockedSet = new Set(riskBlockedGithubIds);
    const approvedGithubIds = scoreableResults.flatMap((result) =>
        result.approved &&
        Number.isInteger(result.github_id) &&
        result.github_id > 0 &&
        !riskBlockedSet.has(result.github_id)
            ? [result.github_id]
            : [],
    );

    summarize(orderedResults);

    if (config.verbose) {
        console.log("\n📊 Score breakdown samples (first 20):");
        for (const result of orderedResults.slice(0, 20)) {
            const flags = result.risk_flags.join(", ");
            const suffix = flags ? ` | risk: ${flags}` : "";
            console.log(
                `   github_id=${result.github_id}: ${Number(result.details?.total ?? 0).toFixed(1)} (${result.reason})${suffix}`,
            );
        }
    }

    if (config.dryRun) {
        if (deletedGithubIds.length > 0) {
            console.log(
                `\n🚫 Dry run would ban ${deletedGithubIds.length} users with deleted GitHub accounts`,
            );
        }
        if (riskBlockedGithubIds.length > 0) {
            console.log(
                `🚩 Dry run would keep ${riskBlockedGithubIds.length} users at spore due to suspicious GitHub profiles`,
            );
        }
        if (unavailableCount > 0) {
            console.log(
                `⏭️ Dry run would defer ${unavailableCount} users because GitHub scoring was unavailable`,
            );
        }
        console.log(
            `🌱 Dry run would upgrade ${approvedGithubIds.length} users to seed`,
        );
        return;
    }

    if (deletedGithubIds.length > 0) {
        const banned = banUsersByGithubIds(config.env, deletedGithubIds);
        console.log(`\n🚫 Banned ${banned} users with deleted GitHub accounts`);
    }

    const stored = storeGithubScores(config.env, "spore", scoreableResults);
    const upgraded = upgradeUsers(config.env, approvedGithubIds);

    console.log("\n📊 Results:");
    console.log(`   Scores stored: ${stored}`);
    console.log(`   Risk-blocked from seed: ${riskBlockedGithubIds.length}`);
    console.log(`   Deferred (GitHub unavailable): ${unavailableCount}`);
    console.log(`   Upgraded to seed: ${upgraded}`);
}

main().catch((error) => {
    console.error(error);
    process.exit(1);
});

File: enter.pollinations.ai/scripts/user-pipeline/daily-spore-recheck.py
#!/usr/bin/env python3
"""Daily spore recheck for seed tier eligibility.

This is the steady-state weekly rotation job:
  - select unbanned spore users
  - order by oldest score_checked_at
  - recheck the oldest ceil(total_spores / 7)
  - persist score and score_checked_at
  - keep suspicious GitHub profiles at spore
  - upgrade qualified non-suspicious users to seed immediately
"""

import argparse
import math
import sys
from datetime import datetime, timezone
from pathlib import Path

SCRIPT_ROOT = Path(__file__).resolve().parent
sys.path.insert(0, str(SCRIPT_ROOT / "shared"))
from python_runtime import ensure_python_bin

ensure_python_bin()

sys.path.insert(0, str(SCRIPT_ROOT / "scoring"))

from d1 import ensure_safe_env, run_d1_query
from github_account_state import (
    D1_BATCH_SIZE,
    GITHUB_USERNAME_RE,
    ban_github_users,
    ban_users_by_emails,
    extract_deleted_github_usernames,
)
from github_score import validate_users

MAX_USERS_PER_RUN = 8000
SQL_BATCH_SIZE = 200
SEED_TIER_BALANCE = 3.0


def load_email_cohort(file_path: str | None) -> list[str] | None:
    if not file_path:
        return None

    try:
        content = Path(file_path).read_text(encoding="utf-8")
    except OSError as error:
        print(f"❌ Failed to read --emails-file {file_path}: {error}", file=sys.stderr)
        sys.exit(1)

    emails = list(
        dict.fromkeys(
            line.strip()
            for line in content.splitlines()
            if line.strip() and not line.strip().startswith("#")
        )
    )
    if not emails:
        print(f"❌ --emails-file {file_path} did not contain any emails.", file=sys.stderr)
        sys.exit(1)
    return emails


def build_email_filter(emails: list[str] | None) -> str:
    if not emails:
        return ""
    values = ", ".join("'" + email.replace("'", "''") + "'" for email in emails)
    return f" AND email IN ({values})"


def fetch_spore_count(env: str, cohort_emails: list[str] | None = None) -> int:
    email_filter = build_email_filter(cohort_emails)
    results = run_d1_query(
        f"""
        SELECT COUNT(*) as count
        FROM user
        WHERE tier = 'spore'
        AND COALESCE(banned, 0) = 0
        {email_filter}
        """,
        env,
    )
    if results is None:
        raise RuntimeError("Failed to fetch current spore count from D1")
    if not results:
        return 0
    return int(results[0]["count"])


def fetch_spore_slice(
    env: str, cohort_emails: list[str] | None = None
) -> tuple[list[dict], int, int]:
    total_spores = fetch_spore_count(env, cohort_emails)
    if total_spores == 0:
        return [], 0, 0

    slice_size = math.ceil(total_spores / 7)
    slice_size = min(slice_size, MAX_USERS_PER_RUN)
    email_filter = build_email_filter(cohort_emails)

    results = run_d1_query(
        f"""
        SELECT email, github_username
        FROM user
        WHERE tier = 'spore'
        AND COALESCE(banned, 0) = 0
        {email_filter}
        ORDER BY score_checked_at ASC, created_at ASC, email ASC
        LIMIT {slice_size}
        """,
        env,
    )
    if results is None:
        raise RuntimeError("Failed to fetch spore recheck slice from D1")
    return results or [], total_spores, slice_size


def store_scores(results: list[dict], env: str) -> tuple[int, int]:
    now = int(datetime.now(timezone.utc).timestamp() * 1000)
    stored = 0
    skipped = 0

    for index in range(0, len(results), SQL_BATCH_SIZE):
        batch = results[index : index + SQL_BATCH_SIZE]
        sanitized_batch = []
        for result in batch:
            username = result.get("username")
            if not isinstance(username, str) or not GITHUB_USERNAME_RE.match(username):
                skipped += 1
                continue

            raw_score = (result.get("details") or {}).get("total", 0)
            total_score = float(raw_score) if raw_score is not None else 0.0
            sanitized_batch.append((username, total_score))

        if not sanitized_batch:
            continue

        score_cases = " ".join(
            f"WHEN '{username}' THEN {score}" for username, score in sanitized_batch
        )
        username_list = ", ".join(
            f"'{username}'" for username, _score in sanitized_batch
        )
        update_query = f"""
            UPDATE user
            SET
                score = CASE github_username {score_cases} END,
                score_checked_at = {now}
            WHERE github_username IN ({username_list})
            AND tier = 'spore'
        """

        update_result = run_d1_query(update_query, env)
        if update_result is None:
            print(
                f"❌ Failed to store batch {index // SQL_BATCH_SIZE + 1}",
                file=sys.stderr,
            )
            continue

        stored += len(sanitized_batch)

    return stored, skipped


def extract_risk_blocked_usernames(results: list[dict]) -> list[str]:
    return list(
        dict.fromkeys(
            result["username"]
            for result in results
            if result.get("risk_status") == "suspicious"
            and isinstance(result.get("username"), str)
        )
    )


def upgrade_users(usernames: list[str], env: str) -> tuple[int, bool]:
    total_upgraded = 0
    failed = False

    for i in range(0, len(usernames), D1_BATCH_SIZE):
        batch = usernames[i : i + D1_BATCH_SIZE]
        safe_batch = [username for username in batch if GITHUB_USERNAME_RE.match(username)]
        if not safe_batch:
            continue

        username_list = ", ".join(f"'{username}'" for username in safe_batch)
        update_query = f"""
            UPDATE user
            SET tier = 'seed', tier_balance = {SEED_TIER_BALANCE}
            WHERE github_username IN ({username_list})
            AND tier = 'spore'
        """
        result = run_d1_query(update_query, env)
        if result is None:
            failed = True
            print(f"   ❌ Batch {i // D1_BATCH_SIZE + 1} failed")
            continue

        total_upgraded += len(safe_batch)

    return total_upgraded, failed


def summarize(results: list[dict]) -> None:
    approved = [result for result in results if result.get("approved")]
    rejected = [result for result in results if not result.get("approved")]
    suspicious = [
        result for result in results if result.get("risk_status") == "suspicious"
    ]

    print("\n📊 Validation summary:")
    print(f"   Approved by score: {len(approved)}")
    print(f"   Rejected by score: {len(rejected)}")
    print(f"   Suspicious GitHub profiles: {len(suspicious)}")

    if results:
        average_score = sum(
            float((result.get("details") or {}).get("total", 0)) for result in results
        ) / len(results)
        print(f"   Average score: {average_score:.2f}")


def main() -> int:
    try:
        parser = argparse.ArgumentParser(description="Daily spore recheck for seed tier")
        parser.add_argument(
            "--dry-run", action="store_true", help="Validate only, no writes"
        )
        parser.add_argument(
            "--env",
            choices=["staging"],
            default="staging",
            help="Environment",
        )
        parser.add_argument(
            "--verbose", "-v", action="store_true", help="Show detailed score breakdowns"
        )
        parser.add_argument(
            "--emails-file",
            help="Only process emails listed in a newline-separated file",
        )
        args = parser.parse_args()

        env = ensure_safe_env(args.env)
        cohort_emails = load_email_cohort(args.emails_file)

        print("🌱 Daily Spore Recheck")
        print(f"   Environment: {env}")
        print(f"   Mode: {'DRY RUN' if args.dry_run else 'LIVE'}")
        if cohort_emails:
            print(f"   Email cohort: {len(cohort_emails)} users")

        rows, total_spores, slice_size = fetch_spore_slice(env, cohort_emails)
        print(f"   Total spores: {total_spores}")
        print(f"   Daily target: {slice_size}")
        print(f"   Selected users: {len(rows)}")

        if not rows:
            print("✅ No spore users to process")
            return 0

        invalid_email_rows = [
            row
            for row in rows
            if not isinstance(row.get("github_username"), str)
            or not GITHUB_USERNAME_RE.match(row["github_username"])
        ]
        valid_usernames = [
            row["github_username"]
            for row in rows
            if isinstance(row.get("github_username"), str)
            and GITHUB_USERNAME_RE.match(row["github_username"])
        ]

        if invalid_email_rows:
            if args.dry_run:
                print(
                    f"🚫 Dry run would ban {len(invalid_email_rows)} spore users with missing/invalid GitHub usernames"
                )
            else:
                banned = ban_users_by_emails(
                    [
                        row["email"]
                        for row in invalid_email_rows
                        if isinstance(row.get("email"), str)
                    ],
                    env,
                )
                print(
                    f"🚫 Banned {banned} spore users with missing/invalid GitHub usernames"
                )

        if not valid_usernames:
            print("✅ No valid spore users left for GitHub scoring")
            return 0

        results = validate_users(valid_usernames)
        results_by_username = {
            result["username"]: result
            for result in results
            if isinstance(result.get("username"), str)
        }
        ordered_results = [
            results_by_username[username]
            for username in valid_usernames
            if username in results_by_username
        ]

        deleted_usernames = extract_deleted_github_usernames(ordered_results)
        deleted_username_set = set(deleted_usernames)
        scoreable_results = [
            result
            for result in ordered_results
            if isinstance(result.get("username"), str)
            and result["username"] not in deleted_username_set
        ]
        risk_blocked_usernames = extract_risk_blocked_usernames(scoreable_results)
        risk_blocked_set = set(risk_blocked_usernames)
        approved_usernames = [
            result["username"]
            for result in scoreable_results
            if result.get("approved")
            and isinstance(result.get("username"), str)
            and result["username"] not in risk_blocked_set
        ]

        summarize(ordered_results)

        if args.verbose:
            print("\n📊 Score breakdown samples (first 20):")
            for result in ordered_results[:20]:
                score = float((result.get("details") or {}).get("total", 0))
                flags = ", ".join(result.get("risk_flags") or [])
                suffix = f" | risk: {flags}" if flags else ""
                print(f"   {result['username']}: {score:.1f} ({result['reason']}){suffix}")

        if args.dry_run:
            if deleted_usernames:
                print(
                    f"\n🚫 Dry run would ban {len(deleted_usernames)} users with deleted/invalid GitHub accounts"
                )
            if risk_blocked_usernames:
                print(
                    f"🚩 Dry run would keep {len(risk_blocked_usernames)} users at spore due to suspicious GitHub profiles"
                )
            print(f"🌱 Dry run would upgrade {len(approved_usernames)} users to seed")
            return 0

        if deleted_usernames:
            banned = ban_github_users(deleted_usernames, env)
            print(f"\n🚫 Banned {banned} users with deleted/invalid GitHub accounts")

        stored, skipped = store_scores(scoreable_results, env)
        upgraded, had_failures = upgrade_users(approved_usernames, env)

        print("\n📊 Results:")
        print(f"   Scores stored: {stored}")
        print(f"   Risk-blocked from seed: {len(risk_blocked_usernames)}")
        print(f"   Upgraded to seed: {upgraded}")
        if skipped:
            print(f"   Skipped invalid usernames: {skipped}")
        if had_failures:
            print("   ❌ Some upgrade batches failed")

        return 1 if had_failures else 0
    except Exception as error:
        print(f"❌ {error}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    sys.exit(main())
